designs/2019-rule-tester-improvements/design.md

```diff
@@ -0,0 +1,83 @@
+- Start Date: 2019-06-07
+- RFC PR: https://github.com/eslint/rfcs/pull/25
+- Authors: Toru Nagashima ([@mysticatea](https://github.com/mysticatea))
+
+# `RuleTester` Improvements
+
+## Summary
+
+This RFC improves `RuleTester` to check more mistakes.
+
+## Motivation
+
+`RuleTester` overlooks some mistakes.
+
+- Using non-standard properties of AST ([typescript-eslint/typescript-eslint#405](https://github.com/typescript-eslint/typescript-eslint/issues/405)).<br>
+  Especially, `node.start` and `node.end` exist in AST `espree` made, but it's not standardized and some custom parsers don't make those properties. But `node.loc` has `start`/`end` properties, so it's hard to detect `node.start`/`node.end` with static analysis. Therefore, `RuleTester` should detect those.
+- Untested autofix.<br>
+  If people forgot to write `output` property in test cases, `RuleTester` doesn't test autofix silently.
+- `errors` property with a number (found in [eslint/eslint#11798](https://github.com/eslint/eslint/pull/11798)).<br>
+  `errors` property with a number ignores syntax errors in test code. We overlooked the mistake of [tests/lib/rules/complexity.js#L84](https://github.com/eslint/eslint/blob/cb1922bdc07e58de0e55c13fd992dd8faf3292a4/tests/lib/rules/complexity.js#L84) due to this. The number `errors` property cannot check the reported error was the expected error.
+
+## Detailed Design
+
+1. Disallowing `node.start` and `node.end`
+1. Ensuring to test autofix
+1. Deprecating the `errors` property with a number
+
+### 1. Disallowing `node.start` and `node.end`
+
+`RuleTester` fails test cases if a rule implementation used `node.start` or `node.end` in the test case.
+
+#### Implementation
+
+- In `RuleTester`, it registers an internal custom parser that wraps `espree` or the parser of `item.parser` to `Linter` object.
+- The internal custom parser fixes the AST that the original parser returned, as like [test-parser.js](https://github.com/eslint/eslint/blob/21f3131aa1636afa8e5c01053e0e870f968425b1/tools/internal-testers/test-parser.js).
+
+### 2. Ensuring to test autofix
+
+`RuleTester` fails test cases if a rule implementation fixed code but `output` property was not defined in the test case.
+
+#### Implementation
+
+- If `output` property didn't exist but the rule fixed the code, `RuleTester` fails the test case as "The rule fixed the code. Please add 'output' property." It's implemented around [lib/rule-tester/rule-tester.js#L594](https://github.com/eslint/eslint/blob/21f3131aa1636afa8e5c01053e0e870f968425b1/lib/rule-tester/rule-tester.js#L594).
+
+### 3. Deprecating the `errors` property with a number
+
+ESLint prints a deprecation warning if `errors` property was used with a number.
```

Comment:
This was added very long time ago per request of one of the members of the team. I don't have a problem deprecating it, but just wanted to provide some context.

Reply:
There is the context around [L19](https://github.com/eslint/rfcs/pull/25/files#diff-97beab911ed47a9b118186f397b9eeccR19).

Reply:
Yes, I understand. I mean that originally when I wrote RuleTester, it didn't include a way to just specify number of errors. It was requested by team member and I added it in. Since it was close to 6 years ago, I don't exactly remember what was the original reasoning, but there was one.

Reply:
[The initial commit](https://github.com/eslint/eslint-tester/blame/master/lib/eslint-tester.js) looks to have that.

I'm guessing that it was to test indeterministic error messages such as including an absolute path to a file. But we can use regular expressions for that purpose now. I want to deprecate `errors` of a number because we cannot check actual errors are the expected errors.

Reply:
Is the concern that some people might be using the `errors` number because they aren't aware of the more strict options? Some other projects don't have as strict standards for rule testing as we have, so it seems like it could be useful to provide a more lax option that makes it easier to write rule tests. (For example, in a project-specific rule, error messages might not be as important because they are only shown to other people working on the project.)

Reply:
Here's the original PR for RuleTester: https://github.com/eslint/eslint/pull/394. RuleTester was first created as part of ESLint but was moved out into it's own package later (Link that @mysticatea provided) and merged back into core again at some point. Back then we had rules that had unspecified errors, so it was added to support those use-cases. I think I'm leaning towards @platinumazure suggestion to keep this format, but count syntax errors as well as rule errors as part of it.

Reply:
Just to clarify, my suggestion is that parse errors always fail a test, period, regardless of the error count (or any details about errors, when longer forms are used).

Reply:
Ah, sorry, I misunderstood your suggestion. I'd be fine with either way.

Reply:
Hmm. OK. I will rewrite this RFC to keep `errors` with a number.

Reply:
Extremely non-descriptive error for inexperienced rule-writers. This conversation is the top result. Consider updating: https://eslint.org/docs/developer-guide/working-with-rules#applying-fixes-1

Reply:
@christiantjl thanks for the suggestion! Can you please open an issue in https://github.com/eslint/eslint repository?